### PR TITLE
fix(machines): warning message for deployed machines

### PR DESCRIPTION
**Pull request by petermakowski**
## Done

- fix warning message deployed machines count

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

#### Tags
- Go to machine list -> tags (`/MAAS/r/tags`)
- Select create a new tag
- Type in some kernel options
- Verify there is no warning message about the count of deployed machines
- Select a tag that has kernel options and some deployed machines from the list
- Click edit
- Edit the kernel options field
- Verify the warning message is displayed with a correct count of deployed machines

#### Machine listing
- Go to machine list
- Select a few groups, and a few individual machines
- Go to take action -> Tag
- Type in a new name in the input field to create a new tag
- Verify machine.count has been called with "tags: deployed" for both selected groups and items and the resulting count displayed in the warning is correct
- Change grouping to status
- Select a group of machines with a "deployed" status
- Select a few groups, and a few individual machines
- Go to take action -> Tag
- Type in a new name in the input field to create a new tag
- Verify the warning message displays the correct count
- Clear the selection
- Select a few machines and a group with a status other than "deployed"
- verify the deployed machine count in the warning message is correct

## Fixes

Fixes: https://github.com/canonical/app-tribe/issues/1417

## Launchpad issue

Related Launchpad maas issue in the form `lp#number`.

## Backports

In general, please propose fixes against _main_ rather than release branches (e.g. 2.7), unless the fix is only applicable for that specific release. Please apply backport labels to the PR (e.g. "Backport 2.7") for the appropriate releases to target.

Only bug and security fixes should be backported, new features should only land in main.

## Screenshots

It could be helpful to provide some screenshots to aid in QAing the change.

### CODE REVIEW

File: src/app/machines/components/MachineHeaderForms/MachineActionFormWrapper/TagForm/AddTagForm/AddTagForm.test.tsx [hunk 1]
```diff
@@ -128,3 +166,68 @@ it("generates a deployed message for multiple machines", async () => {
...
+    <Provider store={store}>
+      <MemoryRouter initialEntries={[{ pathname: "/tags", key: "testKey" }]}>
+        <AddTagForm
+          name="new-tag"
+          onTagCreated={jest.fn()}
+          selectedMachines={selectedMachines}
+        />
+      </MemoryRouter>
+    </Provider>
+  );
```
Comment (ndv99) [inline src/app/machines/components/MachineHeaderForms/MachineActionFormWrapper/TagForm/AddTagForm/AddTagForm.test.tsx:183]:
```suggestion
  renderWithBrowserRouter(
    <AddTagForm
      name="new-tag"
      onTagCreated={jest.fn()}
      selectedMachines={selectedMachines}
    />,
    { route: "/tags", wrapperProps: { store } }
  );
```

Just make sure to change line 30 to the following to ensure this works:

```ts
const mockStore = configureStore<RootState, {}>();
```

And also import `renderWithBrowserRouter` from testing utils

File: src/app/tags/components/KernelOptionsField/KernelOptionsField.test.tsx [hunk 2]
```diff
@@ -101,23 +111,45 @@ it("displays a deployed machines message when updating a tag", async () => {
...
+  const store = mockStore(state);
+  render(
+    <Provider store={store}>
+      <MemoryRouter>
+        <Formik initialValues={{}} onSubmit={jest.fn()}>
+          <KernelOptionsField deployedMachinesCount={1} />
+        </Formik>
+      </MemoryRouter>
+    </Provider>
+  );
```
Comment (ndv99) [inline src/app/tags/components/KernelOptionsField/KernelOptionsField.test.tsx:124]: This could use renderWithMockStore or renderWithBrowserRouter

**Review by ndv99 — APPROVED** [review]
Just a couple small suggestions but looks great over all, up to you if you want to implement them

File: src/app/tags/components/KernelOptionsField/KernelOptionsField.test.tsx (hunk 2 above)
Comment (petermakowski) [reply src/app/tags/components/KernelOptionsField/KernelOptionsField.test.tsx:124]: Sure, good suggestion!

File: src/app/machines/components/MachineHeaderForms/MachineActionFormWrapper/TagForm/AddTagForm/AddTagForm.test.tsx (hunk 1 above)
Comment (petermakowski) [reply src/app/machines/components/MachineHeaderForms/MachineActionFormWrapper/TagForm/AddTagForm/AddTagForm.test.tsx:183]: Sure, good suggestion!

**Comment by petermakowski** [comment]
> QA passes with the steps provided, but it's worth mentioning that if you select every machine (by clicking the select all checkbox), no warning message is displayed, even though deployed machines are selected. Could be one to fix in another PR if that suits

Good catch, I missed this. Fixed.

Also, I really appreciate you went beyond described QA steps. This is very helpful! 🙌

**Pull request merged**